src/execution_algorithms/kfac/kfac_util.cpp

```diff
@@ -430,6 +430,680 @@ void allgather_blocks(
   }
 }
 
+template <El::Device Device>
+void allgather_inverse_matrices_sizes(
+    const std::vector<std::shared_ptr<kfac_block<Device>>>& blocks,
+    El::Matrix<double, El::Device::CPU>& global_buffer,
+    lbann_comm *comm) {
+
+  const size_t num_blocks = blocks.size();
+  global_buffer.Resize(num_blocks, 4);
+  El::Zeros(global_buffer, global_buffer.Height(), global_buffer.Width());
+
+  int iter=0; 
+  for(auto& block : blocks) {
+    const bool is_my_block = (block->get_inverse_proc_rank() == (size_t) comm->get_rank_in_trainer());
+    if(is_my_block) {
+      auto inverse_size = block->get_inverse_matrices_size_vector(comm);
+
+      for (int i=0; i<4; ++i)
+        global_buffer(iter,i) = inverse_size[i];
+    }
+    iter++;
+  }
+
+  comm->allreduce(
+      (El::AbstractMatrix<double>&) global_buffer,
+      comm->get_KFAC_comm());
+}
+
+template <El::Device Device>
+void allgather_inverse_matrices(
+    const std::vector<std::shared_ptr<kfac_block<Device>>>& blocks,
+    El::Matrix<DataType, Device>& global_buffer,
+    lbann_comm *comm) {
+
+  {
+    El::Zeros(global_buffer, global_buffer.Height(), global_buffer.Width());
+    size_t offset = 0;
+    for(auto& block : blocks) {
+      const bool is_my_block = (block->get_inverse_proc_rank() == (size_t) comm->get_rank_in_trainer());
+      if(is_my_block) {
+        offset = block->get_inverse_matrices(global_buffer, offset);
+      }
+      else{
+        offset += block->get_inverse_matrices_size(comm);
+      }
+    }
+  }
+
+
+  comm->allreduce(
+      (El::AbstractMatrix<DataType>&) global_buffer,
+      comm->get_KFAC_comm());
+  {
+    size_t offset = 0;
+    for(auto& block : blocks) {
+      offset = block->set_inverse_matrices(global_buffer, offset, comm);
+    }
+  }
+}
+
+template<>
+void TranslateBetweenGridsVC
+(El::DistMatrix<DataType,El::STAR,El::VC,El::ELEMENT,El::Device::CPU> const& A,
+  El::DistMatrix<DataType,El::STAR,El::VC,El::ELEMENT,El::Device::CPU>& B){
+  LBANN_ERROR("TranslateBetweenGridsVC function is not implemented for CPUs");
+}
+
+template<typename T, El::Device Device>
+void TranslateBetweenGridsVC
+(El::DistMatrix<T,El::STAR,El::VC,El::ELEMENT,Device> const& A,
+  El::DistMatrix<T,El::STAR,El::VC,El::ELEMENT,Device>& B)
```

Comment:
These template definitions should be reversed. The valid implementation is narrowly targeted toward a specific data type and device, so it should be the override instead of the default. This is also more amenable to CPU builds.
```suggestion
namespace {
#ifdef LBANN_HAS_GPU
using KFACDevice = El::Device::GPU;
#else
using KFACDevice = El::Device::CPU;
#endif // LBANN_HAS_GPU
} // namespace <anon>

template<typename T, El::Device Device>
void TranslateBetweenGridsVC
(El::DistMatrix<T,El::STAR,El::VC,El::ELEMENT,Device> const& A,
  El::DistMatrix<T,El::STAR,El::VC,El::ELEMENT,Device>& B){
  LBANN_ERROR("TranslateBetweenGridsVC function is not implemented for this configuration");
}

template<>
void TranslateBetweenGridsVC
(El::DistMatrix<DataType,El::STAR,El::VC,El::ELEMENT,KFACDevice> const& A,
  El::DistMatrix<DataType,El::STAR,El::VC,El::ELEMENT,KFACDevice>& B)
```